src/devices/Pn5180/Pn5180.cs

```diff
@@ -0,0 +1,1496 @@
+            sendData[1] = (byte)(numberValidBitsLastByte == 8 ? 0 : numberValidBitsLastByte);
+            toSend.CopyTo(sendData.Slice(2));
+            LogInfo.Log($"{nameof(SendDataToCard)}: {nameof(sendData)}, {BitConverter.ToString(sendData.ToArray())}", LogLevel.Debug);
+            try
+            {
+                SpiWrite(sendData);
+            }
+            catch (TimeoutException)
+            {
+                LogInfo.Log($"{nameof(SendDataToCard)}: {nameof(TimeoutException)} in {nameof(SpiWrite)}", LogLevel.Debug);
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Read data from a card.
+        /// </summary>
+        /// <param name="toRead">The span of byte to read</param>
+        /// <returns>True if success</returns>
+        /// <remarks>Using this function you'll have to manage yourself the possible low level communication protocol.
+        /// This function write directly to the card all the bytes. Please make sure you'll first load specific radio frequence settings,
+        /// detect a card, select it and then send data</remarks>
+        public bool ReadDataFromCard(Span<byte> toRead)
+        {
+            if (toRead.Length > 508)
+            {
+                throw new ArgumentException($"Data to read can't be larger than 508 bytes");
+            }
+
+            Span<byte> sendData = stackalloc byte[2];
+            sendData[0] = (byte)Command.READ_DATA;
+            sendData[1] = 0x00;
+            LogInfo.Log($"{nameof(ReadDataFromCard)}: {nameof(sendData)}, {BitConverter.ToString(sendData.ToArray())}", LogLevel.Debug);
+            try
+            {
+                SpiWriteRead(sendData, toRead);
+                LogInfo.Log($"{nameof(ReadDataFromCard)}: {nameof(toRead)}, {BitConverter.ToString(toRead.ToArray())}", LogLevel.Debug);
+            }
+            catch (TimeoutException)
+            {
+                LogInfo.Log($"{nameof(ReadDataFromCard)}: {nameof(TimeoutException)} in {nameof(SpiWriteRead)}", LogLevel.Debug);
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Read data from a card.
+        /// </summary>
+        /// <param name="toRead">>The span of byte to read</param>
+        /// <param name="expectedToRead">The expected number of bytes to read</param>
+        /// <returns>True if success. Will return false if the number of bytes to read is not the same as the expected number to read</returns>
+        /// <remarks>Using this function you'll have to manage yourself the possible low level communication protocol.
+        /// This function write directly to the card all the bytes. Please make sure you'll first load specific radio frequence settings,
+        /// detect a card, select it and then send data</remarks>
+        public bool ReadDataFromCard(Span<byte> toRead, int expectedToRead)
```

Comment:
I'd remove expectedToRead and force users to use correct span length since it's very cheap to slice. Also expectedToRead might suggest that you might read less and still success

Reply:
Btw. how would user know how much to read? Usually these APIs are shapped as folowing:

`bool TryReadDataFromCard(Span<byte> toRead, out int bytesWritten)` so that you can pass too long buffer.

Reply:
I will change the ```byte[] ReadAllDataFromCard``` to ```bool ReadDataFromCard(Span<byte> toRead, out int bytesRead)``` which makes more sense.
And I changed to private the ReadData with the expected length. It's used mainly internally.